Code/Framework/AzFramework/AzFramework/Spawnable/Spawnable.cpp

```diff
@@ -6,12 +6,479 @@
+        const EntityAliasList* aliases, const ListTargetSpawanblesCallback& callback) const
+    {
+        AZ_Assert(aliases, "Attempting to visit entity aliases on a spawnable that wasn't locked.");
+        AZStd::unordered_set<AZ::Data::AssetId> spawnableIds;
+        for (const Spawnable::EntityAlias& alias : *aliases)
+        {
+            // If the spawnable id is not valid it means that the alias is referencing the spawnable it's stored on.
+            if (alias.m_spawnable.GetId().IsValid())
+            {
+                auto it = spawnableIds.find(alias.m_spawnable.GetId());
+                if (it == spawnableIds.end())
+                {
+                    callback(alias.m_spawnable);
+                    spawnableIds.emplace(alias.m_spawnable.GetId());
+                }
+            }
+        }
+    }
+
+    void Spawnable::EntityAliasVisitorBase::ListTargetSpawnables(
+        const EntityAliasList* aliases, AZ::Crc32 tag, const ListTargetSpawanblesCallback& callback) const
+    {
+        AZ_Assert(aliases, "Attempting to visit entity aliases on a spawnable that wasn't locked.");
+        AZStd::unordered_set<AZ::Data::AssetId> spawnableIds;
+        for (const Spawnable::EntityAlias& alias : *aliases)
+        {
+            // If the spawnable id is not valid it means that the alias is referencing the spawnable it's stored on.
+            if (alias.m_tag == tag && alias.m_spawnable.GetId().IsValid())
+            {
+                auto it = spawnableIds.find(alias.m_spawnable.GetId());
+                if (it == spawnableIds.end())
+                {
+                    callback(alias.m_spawnable);
+                    spawnableIds.emplace(alias.m_spawnable.GetId());
+                }
+            }
+        }
+    }
+
+
+    //
+    // EntityAliasVisitor
+    //
+
+
+    Spawnable::EntityAliasVisitor::EntityAliasVisitor(Spawnable& owner, EntityAliasList* entityAliasList)
+        : m_owner(owner)
+        , m_entityAliasList(entityAliasList)
+    {
+    }
+
+    Spawnable::EntityAliasVisitor::~EntityAliasVisitor()
+    {
+        if (IsSet())
+        {
+            Optimize();
+
+            AZ_Assert(
+                m_owner.m_shareState == ShareState::ReadWrite, "Attempting to unlock a spawnable that's not in the locked state (%i).",
+                m_owner.m_shareState.load());
+            m_owner.m_shareState = ShareState::NotShared;
+        }
+    }
+
+    Spawnable::EntityAliasVisitor::EntityAliasVisitor(EntityAliasVisitor&& rhs)
+        : m_owner(rhs.m_owner)
+        , m_entityAliasList(rhs.m_entityAliasList)
+    {
+        m_dirty = rhs.m_dirty;
+
+        rhs.m_entityAliasList = nullptr;
+        rhs.m_dirty = false;
+    }
+
+    auto Spawnable::EntityAliasVisitor::operator=(EntityAliasVisitor&& rhs) -> EntityAliasVisitor&
+    {
+        if (this != &rhs)
+        {
+            this->~EntityAliasVisitor();
+            *this = EntityAliasVisitor(rhs.m_owner, rhs.m_entityAliasList);
```

Comment:
I was trying to avoid placement new because that seems to confuse a lot of people, but the alternative doesn't seem to be much better in that regard. Going to replace this with a placement new as that will also avoid some duplicate code.